python/tests/test_stats.py

```diff
@@ -1584,3 +1585,248 @@ def verify(self, ts, reference_sets, focal=None):
+        # these copies are constructed such that bootstrapping a copy is
+        # identical to bootstrapping the original
+        rng = np.random.default_rng(random_seed)
+        for _i in range(num_replicates):
+            block_weights = rng.multinomial(
+                self.num_blocks, [1.0 / self.num_blocks] * self.num_blocks
+            )
+            new_copy = self.copy()
+            new_copy.multiplicity = block_weights[new_copy.node_block]
+            for s in range(new_copy.num_weights):
+                new_copy.cum_weights[:, s] = np.cumsum(
+                    new_copy.node_weights[:, s] * new_copy.multiplicity
+                )
+                new_copy.cum_prob[:, s] = (
+                    new_copy.cum_weights[:, s] / new_copy.cum_weights[-1, s]
+                )
+            yield new_copy
+
+
+def count_leaf_pairs(nd, tr, sample_sets):
+    # counts number of pairs of leaves that have MRCA in node
+    ss = sample_sets
+    wts = np.array([0] * len(ss), dtype="int32")
+    if tr.is_internal(nd):
+        # number of descendants per child across sample sets
+        ch = tr.children(nd)
+        nd = [[len([j for j in tr.leaves(c) if j in s]) for s in ss] for c in ch]
+        # weights are the number of pairs with a TMRCA at node
+        for s in range(len(ss)):
+            for i in range(len(ch) - 1):
+                for j in range(i + 1, len(ch)):
+                    wts[s] += nd[i][s] * nd[j][s]
+    return wts
+
+
+class TestCoalescenceTimeDistribution:
+    # WIP
+
+    def __init__(self, random_seed=1):
+        self.seed_rng = np.random.default_rng(random_seed)
+
+    def naive_pairwise_counts(self, ts, window_breaks, focal=None):
+        # Calculate counts of pairwise coalescence events within
+        # time windows, the stupid way
+
+        if focal is None:
+            focal = []
+            for i in range(ts.num_samples - 1):
+                for j in range(i + 1, ts.num_samples):
+                    focal += [(i, j)]
+        else:
+            assert np.all([len(x) == 2 for x in focal])
+
+        window_breaks = np.sort(np.unique(np.append(window_breaks, [0.0])))
+        assert np.all(window_breaks >= 0.0)
+        duration = window_breaks[1:] - window_breaks[:-1]
+        num_windows = len(duration)
+
+        pairwise_tmrca = np.zeros(0)
+        for pair in focal:
+            i, j = pair
+            pairwise_tmrca = np.append(
+                pairwise_tmrca, np.array([t.tmrca(i, j) for t in ts.trees()])
+            )
+
+        coalesced, _ = np.histogram(pairwise_tmrca, window_breaks)
+        total = len(pairwise_tmrca) - np.append([0], np.cumsum(coalesced))[:num_windows]
+        rate = -np.log(1.0 - coalesced / total) / duration
+
+        return {"coalesced": coalesced, "total": total, "rate": rate}
+
+    def simulate_equilibrium(self, num_chrom, num_samples, recomb_rate=0.0, Ne=10000):
+        # simulate panmictic population of stable size
```

Comment:
Minor note: It looks like maybe you're going for a statistical test here? That's not the sort of thing we have here - instead, we just have "correctness" tests; for that purpose, a small WF simulation is the way to go, since all the weird things happen (with short genome and high enough recomb and mutation rates to get multiple hits of each).

Reply:
Yes, at the moment this class isn't actually testing anything -- it only generates some examples, so that I can more easily check for methodological correctness as I work out the algorithm. My plan was to make proper tests when we're about to merge. The intent is to run these through pytest, to compare the implementation in the library proper against the one defined here in `test_stats.py`, right?

Reply:
Ultimately I was planning on using [the tests for GNN](https://github.com/tskit-dev/tskit/blob/2d513e354267bb2fb13af014e5ad25f9967b4461/python/tests/test_stats.py#L911) as a template.